File: src/main/java/com/joecollins/graphics/screens/generic/RegionalBreakdownScreen.java
package com.joecollins.graphics.screens.generic;

import com.joecollins.bindings.Bindable;
import com.joecollins.bindings.Binding;
import com.joecollins.bindings.BindingReceiver;
import com.joecollins.bindings.IndexedBinding;
import com.joecollins.graphics.components.MultiSummaryFrame;
import com.joecollins.graphics.utils.StandardFont;
import com.joecollins.models.general.Party;
import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.GridLayout;
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.border.EmptyBorder;
import org.apache.commons.lang3.tuple.ImmutablePair;
import org.apache.commons.lang3.tuple.Pair;

public class RegionalBreakdownScreen extends JPanel {

  private RegionalBreakdownScreen(JLabel titleLabel, MultiSummaryFrame multiSummaryFrame) {
    setLayout(new BorderLayout());
    setBackground(Color.WHITE);
    add(titleLabel, BorderLayout.NORTH);

    JPanel panel = new JPanel();
    panel.setBackground(Color.WHITE);
    panel.setBorder(new EmptyBorder(5, 5, 5, 5));
    panel.setLayout(new GridLayout(1, 1));
    panel.add(multiSummaryFrame);
    add(panel, BorderLayout.CENTER);
  }

  public static SeatBuilder seats(
      Binding<String> totalHeaderBinding,
      Binding<Map<Party, Integer>> totalSeatsBinding,
      Binding<Integer> numTotalSeatsBinding,
      Binding<String> titleBinding) {
    return new SeatBuilder(
        totalHeaderBinding, totalSeatsBinding, numTotalSeatsBinding, titleBinding);
  }

  public static SeatDiffBuilder seatsWithDiff(
      Binding<String> totalHeaderBinding,
      Binding<Map<Party, Integer>> totalSeatsBinding,
      Binding<Map<Party, Integer>> seatDiffBinding,
      Binding<Integer> numTotalSeatsBinding,
      Binding<String> titleBinding) {
    return new SeatDiffBuilder(
        totalHeaderBinding, totalSeatsBinding, seatDiffBinding, numTotalSeatsBinding, titleBinding);
  }

  public static SeatPrevBuilder seatsWithPrev(
      Binding<String> totalHeaderBinding,
      Binding<Map<Party, Integer>> totalSeatsBinding,
      Binding<Map<Party, Integer>> prevSeatsBinding,
      Binding<Integer> numTotalSeatsBinding,
      Binding<String> titleBinding) {
    return new SeatPrevBuilder(
        totalHeaderBinding,
        totalSeatsBinding,
        prevSeatsBinding,
        numTotalSeatsBinding,
        titleBinding);
  }

  private interface Entry {
    Binding<String> getHeaderBinding();

    Binding<List<Pair<Color, String>>> getValueBinding();
  }

  private static class BlankEntry implements Entry {

    @Override
    public Binding<String> getHeaderBinding() {
      return Binding.fixedBinding("");
    }

    @Override
    public Binding<List<Pair<Color, String>>> getValueBinding() {
      return Binding.fixedBinding(List.of());
    }
  }

  private static class SeatEntry extends Bindable<SeatEntry, SeatEntry.Property> implements Entry {
    enum Property {
      PARTY_ORDER,
      NAME,
      SEATS,
      TOTAL_SEATS
    }

    protected List<Party> partyOrder = List.of();
    protected String name = "";
    protected Map<Party, Integer> seats = Map.of();
    protected int totalSeats = 0;

    public void setPartyOrder(List<Party> partyOrder) {
      this.partyOrder = partyOrder;
      onPropertyRefreshed(Property.PARTY_ORDER);
    }

    public void setName(String name) {
      this.name = name;
      onPropertyRefreshed(Property.NAME);
    }

    public void setSeats(Map<Party, Integer> seats) {
      this.seats = seats;
      onPropertyRefreshed(Property.SEATS);
    }

    public void setTotalSeats(int totalSeats) {
      this.totalSeats = totalSeats;
      onPropertyRefreshed(Property.TOTAL_SEATS);
    }

    public Binding<String> getHeaderBinding() {
      return Binding.propertyBinding(this, t -> t.name, Property.NAME);
    }

    public Binding<List<Pair<Color, String>>> getValueBinding() {
      return Binding.propertyBinding(
          this,
          t -> {
            List<Pair<Color, String>> ret =
                t.partyOrder.stream().map(t::getPartyLabel).collect(Collectors.toList());
            ret.add(
                ImmutablePair.of(
                    Color.WHITE,
                    seats.values().stream().mapToInt(i -> i).sum() + "/" + totalSeats));
            return ret;
          },
          Property.PARTY_ORDER,
          Property.SEATS,
          Property.TOTAL_SEATS);
    }

    protected ImmutablePair<Color, String> getPartyLabel(Party party) {
      return ImmutablePair.of(party.getColor(), String.valueOf(seats.getOrDefault(party, 0)));
    }
  }

  private static class SeatDiffEntry extends SeatEntry {
    private static final DecimalFormat DIFF_FORMAT = new DecimalFormat("+0;-0");
    private Map<Party, Integer> diff = Map.of();

    public void setDiff(Map<Party, Integer> diff) {
      this.diff = diff;
      onPropertyRefreshed(Property.SEATS);
    }

    @Override
    protected ImmutablePair<Color, String> getPartyLabel(Party party) {
      int seats = this.seats.getOrDefault(party, 0);
      int diff = this.diff.getOrDefault(party, 0);
      return ImmutablePair.of(
          party.getColor(),
          seats + " (" + (diff == 0 ? "\u00b10" : DIFF_FORMAT.format(diff)) + ")");
    }
  }

  private static class SeatPrevEntry extends SeatEntry {
    private static final DecimalFormat DIFF_FORMAT = new DecimalFormat("+0;-0");
    private Map<Party, Integer> prev = Map.of();

    public void setPrev(Map<Party, Integer> prev) {
      this.prev = prev;
      onPropertyRefreshed(Property.SEATS);
    }

    @Override
    protected ImmutablePair<Color, String> getPartyLabel(Party party) {
      int seats = this.seats.getOrDefault(party, 0);
      int diff = seats - this.prev.getOrDefault(party, 0);
      return ImmutablePair.of(
          party.getColor(),
          seats + " (" + (diff == 0 ? "\u00b10" : DIFF_FORMAT.format(diff)) + ")");
    }
  }

  private static class MultiPartyResultBuilder {
    protected BindingReceiver<String> title;
    protected List<Entry> entries = new ArrayList<>();
    protected BindingReceiver<List<Party>> partyOrder;

    public RegionalBreakdownScreen build(Binding<String> titleBinding) {
      JLabel headerLabel = new JLabel();
      headerLabel.setFont(StandardFont.readBoldFont(32));
      headerLabel.setHorizontalAlignment(JLabel.CENTER);
      headerLabel.setBorder(new EmptyBorder(5, 0, -5, 0));
      titleBinding.bindLegacy(headerLabel::setText);

      return new RegionalBreakdownScreen(headerLabel, createFrame());
    }

    private MultiSummaryFrame createFrame() {
      var frame = new MultiSummaryFrame();
      frame.setHeaderBinding(title.getBinding());
      frame.setNumRowsBinding(Binding.fixedBinding(entries.size()));
      frame.setRowHeaderBinding(IndexedBinding.listBinding(entries, Entry::getHeaderBinding));
      frame.setValuesBinding(IndexedBinding.listBinding(entries, Entry::getValueBinding));
      return frame;
    }
  }

  public static class SeatBuilder extends MultiPartyResultBuilder {

    public SeatBuilder(
        Binding<String> totalHeaderBinding,
        Binding<Map<Party, Integer>> totalSeatsBinding,
        Binding<Integer> numTotalSeatsBinding,
        Binding<String> titleBinding) {
      var totalSeats = new BindingReceiver<>(totalSeatsBinding);
      this.title = new BindingReceiver<>(titleBinding);

      this.partyOrder =
          new BindingReceiver<>(totalSeats.getBinding(RegionalBreakdownScreen::extractPartyOrder));
      SeatEntry topEntry = new SeatEntry();
      partyOrder.getBinding().bindLegacy(topEntry::setPartyOrder);
      totalHeaderBinding.bindLegacy(topEntry::setName);
      totalSeats.getBinding().bindLegacy(topEntry::setSeats);
      numTotalSeatsBinding.bindLegacy(topEntry::setTotalSeats);
      entries.add(topEntry);
    }

    public SeatBuilder withBlankRow() {
      entries.add(new BlankEntry());
      return this;
    }

    public SeatBuilder withRegion(
        Binding<String> nameBinding,
        Binding<Map<Party, Integer>> seatsBinding,
        Binding<Integer> numSeatsBinding) {
      SeatEntry newEntry = new SeatEntry();
      partyOrder.getBinding().bindLegacy(newEntry::setPartyOrder);
      nameBinding.bindLegacy(newEntry::setName);
      seatsBinding.bindLegacy(newEntry::setSeats);
      numSeatsBinding.bindLegacy(newEntry::setTotalSeats);
      entries.add(newEntry);
      return this;
    }
  }

  public static class SeatDiffBuilder extends MultiPartyResultBuilder {

    public SeatDiffBuilder(
        Binding<String> totalHeaderBinding,
        Binding<Map<Party, Integer>> totalSeatsBinding,
        Binding<Map<Party, Integer>> seatDiffBinding,
        Binding<Integer> numTotalSeatsBinding,
        Binding<String> titleBinding) {
      var totalSeats = new BindingReceiver<>(totalSeatsBinding);
      var seatDiff = new BindingReceiver<>(seatDiffBinding);
      this.title = new BindingReceiver<>(titleBinding);

      this.partyOrder =
          new BindingReceiver<>(
              totalSeats
                  .getBinding()
                  .merge(seatDiff.getBinding(), RegionalBreakdownScreen::extractPartyOrder));
      SeatDiffEntry topEntry = new SeatDiffEntry();
      partyOrder.getBinding().bindLegacy(topEntry::setPartyOrder);
      totalHeaderBinding.bindLegacy(topEntry::setName);
      totalSeats.getBinding().bindLegacy(topEntry::setSeats);
      seatDiff.getBinding().bindLegacy(topEntry::setDiff);
      numTotalSeatsBinding.bindLegacy(topEntry::setTotalSeats);
      entries.add(topEntry);
    }

    public SeatDiffBuilder withBlankRow() {
      entries.add(new BlankEntry());
      return this;
    }

    public SeatDiffBuilder withRegion(
        Binding<String> nameBinding,
        Binding<Map<Party, Integer>> seatsBinding,
        Binding<Map<Party, Integer>> diffBinding,
        Binding<Integer> numSeatsBinding) {
      SeatDiffEntry newEntry = new SeatDiffEntry();
      partyOrder.getBinding().bindLegacy(newEntry::setPartyOrder);
      nameBinding.bindLegacy(newEntry::setName);
      seatsBinding.bindLegacy(newEntry::setSeats);
      diffBinding.bindLegacy(newEntry::setDiff);
      numSeatsBinding.bindLegacy(newEntry::setTotalSeats);
      entries.add(newEntry);
      return this;
    }
  }

  public static class SeatPrevBuilder extends MultiPartyResultBuilder {

    public SeatPrevBuilder(
        Binding<String> totalHeaderBinding,
        Binding<Map<Party, Integer>> totalSeatsBinding,
        Binding<Map<Party, Integer>> prevSeatBinding,
        Binding<Integer> numTotalSeatsBinding,
        Binding<String> titleBinding) {
      var totalSeats = new BindingReceiver<>(totalSeatsBinding);
      var prevSeats = new BindingReceiver<>(prevSeatBinding);
      this.title = new BindingReceiver<>(titleBinding);

      this.partyOrder =
          new BindingReceiver<>(
              totalSeats
                  .getBinding()
                  .merge(prevSeats.getBinding(), RegionalBreakdownScreen::extractPartyOrder));
      SeatPrevEntry topEntry = new SeatPrevEntry();
      partyOrder.getBinding().bindLegacy(topEntry::setPartyOrder);
      totalHeaderBinding.bindLegacy(topEntry::setName);
      totalSeats.getBinding().bindLegacy(topEntry::setSeats);
      prevSeats.getBinding().bindLegacy(topEntry::setPrev);
      numTotalSeatsBinding.bindLegacy(topEntry::setTotalSeats);
      entries.add(topEntry);
    }

    public SeatPrevBuilder withBlankRow() {
      entries.add(new BlankEntry());
      return this;
    }

    public SeatPrevBuilder withRegion(
        Binding<String> nameBinding,
        Binding<Map<Party, Integer>> seatsBinding,
        Binding<Map<Party, Integer>> prevBinding,
        Binding<Integer> numSeatsBinding) {
      SeatPrevEntry newEntry = new SeatPrevEntry();
      partyOrder.getBinding().bindLegacy(newEntry::setPartyOrder);
      nameBinding.bindLegacy(newEntry::setName);
      seatsBinding.bindLegacy(newEntry::setSeats);
      prevBinding.bindLegacy(newEntry::setPrev);
      numSeatsBinding.bindLegacy(newEntry::setTotalSeats);
      entries.add(newEntry);
      return this;
    }
  }

  private static List<Party> extractPartyOrder(Map<Party, Integer> result) {
    return result.entrySet().stream()
        .filter(e -> e.getValue() > 0)
        .sorted(Map.Entry.<Party, Integer>comparingByValue().reversed())
        .map(Map.Entry::getKey)
        .collect(Collectors.toList());
  }

  private static List<Party> extractPartyOrder(
      Map<Party, Integer> result, Map<Party, Integer> diff) {
    return Stream.concat(result.keySet().stream(), diff.keySet().stream())
        .distinct()
        .filter(p -> result.getOrDefault(p, 0) > 0 || diff.getOrDefault(p, 0) != 0)
        .sorted(Comparator.comparing(p -> result.getOrDefault(p, 0)).reversed())
        .collect(Collectors.toList());
  }
}

File: src/main/java/com/joecollins/graphics/screens/generic/RegionalBreakdownScreen.kt
package com.joecollins.graphics.screens.generic

import com.joecollins.bindings.Bindable
import com.joecollins.bindings.Binding
import com.joecollins.bindings.BindingReceiver
import com.joecollins.bindings.IndexedBinding
import com.joecollins.graphics.components.MultiSummaryFrame
import com.joecollins.graphics.utils.StandardFont
import com.joecollins.models.general.Party
import java.awt.BorderLayout
import java.awt.Color
import java.awt.GridLayout
import java.text.DecimalFormat
import java.util.ArrayList
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.border.EmptyBorder
import org.apache.commons.lang3.tuple.ImmutablePair
import org.apache.commons.lang3.tuple.Pair

class RegionalBreakdownScreen private constructor(titleLabel: JLabel, multiSummaryFrame: MultiSummaryFrame) : JPanel() {
    interface Entry {
        val headerBinding: Binding<String>
        val valueBinding: Binding<List<Pair<Color, String>>>
    }

    private class BlankEntry : Entry {
        override val headerBinding: Binding<String>
            get() = Binding.fixedBinding("")
        override val valueBinding: Binding<List<Pair<Color, String>>>
            get() = Binding.fixedBinding(listOf())
    }

    private open class SeatEntry : Bindable<SeatEntry, SeatEntry.Property>(), Entry {
        enum class Property {
            PARTY_ORDER, NAME, SEATS, TOTAL_SEATS
        }

        private var _partyOrder: List<Party> = emptyList()
        private var _name = ""
        private var _seats: Map<Party, Int> = emptyMap()
        private var _totalSeats = 0

        var partyOrder: List<Party>
        get() = _partyOrder
        set(partyOrder) {
            _partyOrder = partyOrder
            onPropertyRefreshed(Property.PARTY_ORDER)
        }

        var name: String
        get() = _name
        set(name) {
            _name = name
            onPropertyRefreshed(Property.NAME)
        }

        var seats: Map<Party, Int>
        get() = _seats
        set(seats) {
            _seats = seats
            onPropertyRefreshed(Property.SEATS)
        }

        var totalSeats: Int
        get() = _totalSeats
        set(totalSeats) {
            _totalSeats = totalSeats
            onPropertyRefreshed(Property.TOTAL_SEATS)
        }

        override val headerBinding: Binding<String>
            get() = Binding.propertyBinding(this, { t: SeatEntry -> t.name }, Property.NAME)
        override val valueBinding: Binding<List<Pair<Color, String>>>
            get() = Binding.propertyBinding(
                    this,
                    { t: SeatEntry ->
                        val ret: MutableList<Pair<Color, String>> = t.partyOrder.map { t.getPartyLabel(it) }.toMutableList()
                        ret.add(
                                ImmutablePair.of(
                                        Color.WHITE,
                                        seats.values.sum().toString() + "/" + totalSeats))
                        ret
                    },
                    Property.PARTY_ORDER,
                    Property.SEATS,
                    Property.TOTAL_SEATS)

        protected open fun getPartyLabel(party: Party): Pair<Color, String> {
            return ImmutablePair.of(party.color, (seats[party] ?: 0).toString())
        }
    }

    private class SeatDiffEntry : SeatEntry() {
        private var _diff: Map<Party, Int> = java.util.Map.of()

        var diff: Map<Party, Int>
        get() = _diff
        set(diff) {
            _diff = diff
            onPropertyRefreshed(Property.SEATS)
        }

        override fun getPartyLabel(party: Party): ImmutablePair<Color, String> {
            val seats = seats[party] ?: 0
            val diff = diff[party] ?: 0
            return ImmutablePair.of(
                    party.color,
                    seats.toString() + " (" + (if (diff == 0) "\u00b10" else DIFF_FORMAT.format(diff.toLong())) + ")")
        }

        companion object {
            private val DIFF_FORMAT = DecimalFormat("+0;-0")
        }
    }

    private class SeatPrevEntry : SeatEntry() {
        private var _prev: Map<Party, Int> = java.util.Map.of()

        var prev: Map<Party, Int>
        get() = _prev
        set(prev) {
            _prev = prev
            onPropertyRefreshed(Property.SEATS)
        }

        override fun getPartyLabel(party: Party): ImmutablePair<Color, String> {
            val seats = seats[party] ?: 0
            val diff = seats - (prev[party] ?: 0)
            return ImmutablePair.of(
                    party.color,
                    seats.toString() + " (" + (if (diff == 0) "\u00b10" else DIFF_FORMAT.format(diff.toLong())) + ")")
        }

        companion object {
            private val DIFF_FORMAT = DecimalFormat("+0;-0")
        }
    }

    open class MultiPartyResultBuilder(
        titleBinding: Binding<String>
    ) {
        protected val title: BindingReceiver<String> = BindingReceiver(titleBinding)
        protected val entries: MutableList<Entry> = ArrayList()
        protected var partyOrder: BindingReceiver<List<Party>>? = null

        fun build(titleBinding: Binding<String?>): RegionalBreakdownScreen {
            val headerLabel = JLabel()
            headerLabel.font = StandardFont.readBoldFont(32)
            headerLabel.horizontalAlignment = JLabel.CENTER
            headerLabel.border = EmptyBorder(5, 0, -5, 0)
            titleBinding.bind { headerLabel.text = it }
            return RegionalBreakdownScreen(headerLabel, createFrame())
        }

        private fun createFrame(): MultiSummaryFrame {
            val frame = MultiSummaryFrame()
            frame.setHeaderBinding(title.getBinding())
            frame.setNumRowsBinding(Binding.fixedBinding(entries.size))
            frame.setRowHeaderBinding(IndexedBinding.listBinding(entries) { obj: Entry -> obj.headerBinding })
            frame.setValuesBinding(IndexedBinding.listBinding(entries) { obj: Entry -> obj.valueBinding })
            return frame
        }
    }

    class SeatBuilder(
        totalHeaderBinding: Binding<String>,
        totalSeatsBinding: Binding<Map<Party, Int>>,
        numTotalSeatsBinding: Binding<Int>,
        titleBinding: Binding<String>
    ) : MultiPartyResultBuilder(titleBinding) {

        fun withBlankRow(): SeatBuilder {
            entries.add(BlankEntry())
            return this
        }

        fun withRegion(
            nameBinding: Binding<String>,
            seatsBinding: Binding<Map<Party, Int>>,
            numSeatsBinding: Binding<Int>
        ): SeatBuilder {
            val newEntry = SeatEntry()
            partyOrder!!.getBinding().bind { newEntry.partyOrder = it }
            nameBinding.bind { newEntry.name = it }
            seatsBinding.bind { newEntry.seats = it }
            numSeatsBinding.bind { newEntry.totalSeats = it }
            entries.add(newEntry)
            return this
        }

        init {
            val totalSeats = BindingReceiver(totalSeatsBinding)
            partyOrder = BindingReceiver(totalSeats.getBinding { result: Map<Party, Int> -> extractPartyOrder(result) })
            val topEntry = SeatEntry()
            partyOrder!!.getBinding().bind { topEntry.partyOrder = it }
            totalHeaderBinding.bind { topEntry.name = it }
            totalSeats.getBinding().bind { topEntry.seats = it }
            numTotalSeatsBinding.bind { topEntry.totalSeats = it }
            entries.add(topEntry)
        }
    }

    class SeatDiffBuilder(
        totalHeaderBinding: Binding<String>,
        totalSeatsBinding: Binding<Map<Party, Int>>,
        seatDiffBinding: Binding<Map<Party, Int>>,
        numTotalSeatsBinding: Binding<Int>,
        titleBinding: Binding<String>
    ) : MultiPartyResultBuilder(titleBinding) {
        fun withBlankRow(): SeatDiffBuilder {
            entries.add(BlankEntry())
            return this
        }

        fun withRegion(
            nameBinding: Binding<String>,
            seatsBinding: Binding<Map<Party, Int>>,
            diffBinding: Binding<Map<Party, Int>>,
            numSeatsBinding: Binding<Int>
        ): SeatDiffBuilder {
            val newEntry = SeatDiffEntry()
            partyOrder!!.getBinding().bind { newEntry.partyOrder = it }
            nameBinding.bind { newEntry.name = it }
            seatsBinding.bind { newEntry.seats = it }
            diffBinding.bind { newEntry.diff = it }
            numSeatsBinding.bind { newEntry.totalSeats = it }
            entries.add(newEntry)
            return this
        }

        init {
            val totalSeats = BindingReceiver(totalSeatsBinding)
            val seatDiff = BindingReceiver(seatDiffBinding)
            partyOrder = BindingReceiver(
                    totalSeats
                            .getBinding()
                            .merge(seatDiff.getBinding()) { result, diff -> extractPartyOrder(result, diff) })
            val topEntry = SeatDiffEntry()
            partyOrder!!.getBinding().bind { topEntry.partyOrder = it }
            totalHeaderBinding.bind { topEntry.name = it }
            totalSeats.getBinding().bind { topEntry.seats = it }
            seatDiff.getBinding().bind { topEntry.diff = it }
            numTotalSeatsBinding.bind { topEntry.totalSeats = it }
            entries.add(topEntry)
        }
    }

    class SeatPrevBuilder(
        totalHeaderBinding: Binding<String>,
        totalSeatsBinding: Binding<Map<Party, Int>>,
        prevSeatBinding: Binding<Map<Party, Int>>,
        numTotalSeatsBinding: Binding<Int>,
        titleBinding: Binding<String>
    ) : MultiPartyResultBuilder(titleBinding) {

        fun withBlankRow(): SeatPrevBuilder {
            entries.add(BlankEntry())
            return this
        }

        fun withRegion(
            nameBinding: Binding<String>,
            seatsBinding: Binding<Map<Party, Int>>,
            prevBinding: Binding<Map<Party, Int>>,
            numSeatsBinding: Binding<Int>
        ): SeatPrevBuilder {
            val newEntry = SeatPrevEntry()
            partyOrder!!.getBinding().bind { newEntry.partyOrder = it }
            nameBinding.bind { newEntry.name = it }
            seatsBinding.bind { newEntry.seats = it }
            prevBinding.bind { newEntry.prev = it }
            numSeatsBinding.bind { newEntry.totalSeats = it }
            entries.add(newEntry)
            return this
        }

        init {
            val totalSeats = BindingReceiver(totalSeatsBinding)
            val prevSeats = BindingReceiver(prevSeatBinding)
            partyOrder = BindingReceiver(
                    totalSeats
                            .getBinding()
                            .merge(prevSeats.getBinding()) { result: Map<Party, Int>, diff: Map<Party, Int> -> extractPartyOrder(result, diff) })
            val topEntry = SeatPrevEntry()
            partyOrder!!.getBinding().bind { topEntry.partyOrder = it }
            totalHeaderBinding.bind { topEntry.name = it }
            totalSeats.getBinding().bind { topEntry.seats = it }
            prevSeats.getBinding().bind { topEntry.prev = it }
            numTotalSeatsBinding.bind { topEntry.totalSeats = it }
            entries.add(topEntry)
        }
    }

    companion object {
        @JvmStatic fun seats(
            totalHeaderBinding: Binding<String>,
            totalSeatsBinding: Binding<Map<Party, Int>>,
            numTotalSeatsBinding: Binding<Int>,
            titleBinding: Binding<String>
        ): SeatBuilder {
            return SeatBuilder(
                    totalHeaderBinding, totalSeatsBinding, numTotalSeatsBinding, titleBinding)
        }

        @JvmStatic fun seatsWithDiff(
            totalHeaderBinding: Binding<String>,
            totalSeatsBinding: Binding<Map<Party, Int>>,
            seatDiffBinding: Binding<Map<Party, Int>>,
            numTotalSeatsBinding: Binding<Int>,
            titleBinding: Binding<String>
        ): SeatDiffBuilder {
            return SeatDiffBuilder(
                    totalHeaderBinding, totalSeatsBinding, seatDiffBinding, numTotalSeatsBinding, titleBinding)
        }

        @JvmStatic fun seatsWithPrev(
            totalHeaderBinding: Binding<String>,
            totalSeatsBinding: Binding<Map<Party, Int>>,
            prevSeatsBinding: Binding<Map<Party, Int>>,
            numTotalSeatsBinding: Binding<Int>,
            titleBinding: Binding<String>
        ): SeatPrevBuilder {
            return SeatPrevBuilder(
                    totalHeaderBinding,
                    totalSeatsBinding,
                    prevSeatsBinding,
                    numTotalSeatsBinding,
                    titleBinding)
        }

        private fun extractPartyOrder(result: Map<Party, Int>): List<Party> {
            return result.entries.asSequence()
                    .filter { it.value > 0 }
                    .sortedByDescending { it.value }
                    .map { it.key }
                    .toList()
        }

        private fun extractPartyOrder(
            result: Map<Party, Int>,
            diff: Map<Party, Int>
        ): List<Party> {
            return sequenceOf(result.keys.asSequence(), diff.keys.asSequence()).flatten()
                    .distinct()
                    .filter { party -> result[party] ?: 0 > 0 || diff[party] ?: 0 != 0 }
                    .sortedByDescending { party -> result[party] ?: 0 }
                    .toList()
        }
    }

    init {
        layout = BorderLayout()
        background = Color.WHITE
        add(titleLabel, BorderLayout.NORTH)
        val panel = JPanel()
        panel.background = Color.WHITE
        panel.border = EmptyBorder(5, 5, 5, 5)
        panel.layout = GridLayout(1, 1)
        panel.add(multiSummaryFrame)
        add(panel, BorderLayout.CENTER)
    }
}
